base/stat.jl

```diff
@@ -464,22 +464,16 @@ end
 islink(path...) = islink(lstat(path...))
 
 # samefile can be used for files and directories: #11145#issuecomment-99511194
-samefile(a::StatStruct, b::StatStruct) = a.device==b.device && a.inode==b.inode
+function samefile(a::StatStruct, b::StatStruct)
+    ispath(a) && ispath(b) && a.device == b.device && a.inode == b.inode
+end
 
 """
     samefile(path_a::AbstractString, path_b::AbstractString)
 
 Check if the paths `path_a` and `path_b` refer to the same existing file or directory.
 """
-function samefile(a::AbstractString, b::AbstractString)
-    infoa = stat(a)
-    infob = stat(b)
-    if ispath(infoa) && ispath(infob)
-        samefile(infoa, infob)
-    else
-        return false
-    end
-end
+samefile(a::AbstractString, b::AbstractString) = samefile(stat(a), stat(b))
```

Comment:
if we want to go beyond just being a bugfix, I propose that we also accept that anytime you give the exact same string, we consider that to be the same file, even if the named file does not exist:
```suggestion
function samefile(a::AbstractString, b::AbstractString)
    a == b && return true
    Sys.iswindows() && joinpath(splitpath(a)) == joinpath(splitpath(b)) && return true
    return samefile(stat(a), stat(b))
end
```
Though I don't know how far to take that, since if neither `stat(a)` or `stat(b)` exist, we could also find the first component that does exist and do a string comparison beyond that point:
```julia
function samefile(a::AbstractString, b::AbstractString)
    a == b && return true
    sa, sb = stat(a), stat(b)
    while !ispath(a) && !ispath(b)
        basename(a) == basename(b) || return false
        a, b = dirname(a), dirname(b)
        sa, sb = stat(a), stat(b)
    end
    return samefile(sa, sb)
end
```
assuming all paths are readable (so `stat` does not error)

or we could just do the path splitting once (which would be simliar a legacy mode extension for `realpath`, albeit one which we do not support).

Reply:
That would differ from the currently documented behavior that the paths must refer to the same _existing file_.

Reply:
yeah, I guess this should be something like `samefilepath` instead, if someone eventually ever needs it